### PR TITLE
v1.8 backports 2021-04-09

### PR DESCRIPTION
**Pull request by errordeveloper**
* #15453 -- eni: Assign primary IP to support multiple VPC CIDRs (@gandro)
 * #15261 -- node-neigh: Reduce arping related log msg's level (@brb)
 * #15598 -- Make encryption+chaining limitations clearer (@joestringer)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 15453 15261 15598; do contrib/backporting/set-labels.py $pr done 1.8; done
```

### CODE REVIEW

File: bpf/Makefile [hunk 1]
```diff
@@ -78,9 +78,16 @@ MAX_BASE_OPTIONS = -DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 \
...
 ifeq ("$(KERNEL)","49")
 # IPSec is incompatible with BPF NodePort so we only enable on 4.9.
-MAX_BASE_OPTIONS += -DENABLE_IPSEC=1
+MAX_BASE_OPTIONS += -DENABLE_IPSEC=1 -DNEEDS_RELAX_VERIFIER=1
 else
-MAX_BASE_OPTIONS += -DHAVE_LPM_TRIE_MAP_TYPE=1 -DHAVE_LRU_HASH_MAP_TYPE=1
+MAX_BASE_OPTIONS += -DHAVE_LPM_TRIE_MAP_TYPE=1 -DHAVE_LRU_HASH_MAP_TYPE=1 \
+	-DENABLE_MASQUERADE=1
+ifeq ("$(KERNEL)","netnext")
+MAX_BASE_OPTIONS += -DENABLE_TPROXY=1 -DENABLE_REDIRECT_FAST=1 -DENABLE_BANDWIDTH_MANAGER=1 \
```
Comment (pchaigno) [inline bpf/Makefile]: I don't think this is the correct conflict resolution as these three datapath options don't exist in v1.8.

File: bpf/Makefile [hunk 2]
```diff
@@ -75,13 +75,18 @@ MAX_BASE_OPTIONS = -DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 \
...
 # IPSec is incompatible with BPF NodePort so we only enable on 4.9.
 MAX_BASE_OPTIONS += -DENABLE_IPSEC=1 -DNEEDS_RELAX_VERIFIER=1
 else
 MAX_BASE_OPTIONS += -DHAVE_LPM_TRIE_MAP_TYPE=1 -DHAVE_LRU_HASH_MAP_TYPE=1 \
-	-DENABLE_MASQUERADE=1
+	-DENABLE_MASQUERADE=1 -DENABLE_SRC_RANGE_CHECK=1 -DENABLE_NODEPORT=1 \
+	-DENABLE_EXTERNAL_IP=1 -DENABLE_NODEPORT_ACCELERATION=1 \
+	-DENABLE_SESSION_AFFINITY=1 -DENABLE_DSR_ICMP_ERRORS=1 -DENABLE_DSR=1 \
+	-DENABLE_HOSTPORT=1 -DENABLE_LOADBALANCER -DENABLE_DSR_HYBRID=1 \
+	-DENABLE_DSR_ICMP_ERRORS=1 -DENABLE_IPV4_FRAGMENTS=1 -DENABLE_CUSTOM_CALLS=1 \
```
Comment (pchaigno) [inline bpf/Makefile]: `ENABLE_CUSTOM_CALLS` and `ENABLE_DSR_ICMP_ERRORS` are v1.10 features.

File: bpf/Makefile [hunk 3]
```diff
@@ -143,11 +144,9 @@ HOST_OPTIONS = $(LXC_OPTIONS) \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DPOLICY_VERDICT_NOTIFY:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_HOST_FIREWALL: \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DPOLICY_VERDICT_NOTIFY:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_SESSION_AFFINITY:-DENABLE_HOST_FIREWALL:
 
-MAX_HOST_OPTIONS = $(MAX_BASE_OPTIONS) -DENCAP_IFINDEX=1 -DENABLE_HOST_FIREWALL=1
+MAX_HOST_OPTIONS = $(MAX_BASE_OPTIONS) -DENCAP_IFINDEX=1
 ifneq ("$(KERNEL)","49")
-MAX_HOST_OPTIONS += -DENABLE_HOST_SERVICES_UDP=1 -DENABLE_HOST_SERVICES_TCP=1 \
-	-DENABLE_NODEPORT=1 -DENABLE_NODEPORT_ACCELERATION=1 -DENABLE_EXTERNAL_IP=1 -DENABLE_DSR=1 \
-	-DENABLE_DSR_HYBRID=1 -DENABLE_SESSION_AFFINITY=1
+MAX_HOST_OPTIONS += -DENABLE_EGRESS_GATEWAY=1
```
Comment (pchaigno) [inline bpf/Makefile]: This is a v1.10 feature.

File: bpf/Makefile [hunk 4]
```diff
@@ -96,6 +95,9 @@ endif
 endif
 
 MAX_LB_OPTIONS = $(MAX_BASE_OPTIONS) -DENABLE_NAT46=1 -DENCAP_IFINDEX=1
+ifneq ("$(KERNEL)","49")
+MAX_LB_OPTIONS += -DLB_SELECTION=1 -DLB_SELECTION_MAGLEV=1
```
Comment (pchaigno) [inline bpf/Makefile]: The `bpf/Makefile: Disable Maglev for bpf_host and bpf_lxc` shouldn't be backported to v1.8 as Maglev was introduced in v1.9.

**Review by pchaigno — DISMISSED** [review]
Backporting https://github.com/cilium/cilium/pull/15540 is a bit tricky (need to know which datapath options existed in v1.8), so I'm happy to take care of that backport if you want.

**Comment by errordeveloper** [comment]
@pchaigno I see, I very opportunistically resolved merged conflicts, but wasn't sure of the underlying changes. I'll drop it from here and the 1.9 backport, please open separate PRs :)

**Comment by errordeveloper** [comment]
test-backport-1.8

File: Documentation/gettingstarted/cni-chaining-azure-cni.rst [hunk 5]
```diff
@@ -8,12 +8,71 @@
...
+   the Azure CNI in a chaining configuration.
+
+.. include:: cni-chaining-limitations.rst
+
+This guide explains how to set up Cilium in combination with Azure CNI in a
+chaining configuration. In this hybrid mode, the Azure CNI plugin is
+responsible for setting up the virtual network devices as well as address
+allocation (IPAM). After the initial networking is setup, the Cilium CNI plugin
+is called to attach eBPF programs to the network devices set up by Azure CNI to
+enforce network policies, perform load-balancing, and encryption.
```
Comment (joestringer) [inline Documentation/gettingstarted/cni-chaining-azure-cni.rst]: Some of these changes do not match the original commits. This workflow for Azure might happen to work but we never included or tested such instructions for v1.8 versions as far as I know. Have you tried these out?

Comment (joestringer) [reply Documentation/gettingstarted/cni-chaining-azure-cni.rst]: Maybe not this particular paragraph, but the bits below.

Comment (errordeveloper) [reply Documentation/gettingstarted/cni-chaining-azure-cni.rst]: I just applied the patch, didn't look at the context very much...

Comment (joestringer) [reply Documentation/gettingstarted/cni-chaining-azure-cni.rst]: Seems like there was maybe a conflict if this text was inadvertently introduced? In such cases it's also good to highlight in the backport PR description message that there was a conflict, so the original authors of those commits can know to take a closer look.

For this case, I'd suggest just doing a `git show` for the original commit (as pointed out in the backport commit message), and just check that the diff looks the same as the backport commit and remove any bits that don't match. Alternatively if you're up for it, you could try running through the full instructions here to make sure that the full changes work.

Comment (pchaigno) [reply Documentation/gettingstarted/cni-chaining-azure-cni.rst]: > In such cases it's also good to highlight in the backport PR description message that there was a conflict, so the original authors of those commits can know to take a closer look.

IMO, all authors of backported commits with conflicts should be requested as reviewers. We would probably need to make that clearer in the docs.

Comment (errordeveloper) [reply Documentation/gettingstarted/cni-chaining-azure-cni.rst]: I thought mention in the PR description was sufficient to get author's attention generally...

**Review by brb — CHANGES_REQUESTED** [review]
My changes depend on https://github.com/cilium/cilium/pull/15537 to be backported first.

File: pkg/aws/ec2/ec2.go [hunk 6]
```diff
@@ -195,8 +195,8 @@ func parseENI(iface *ec2.NetworkInterface, vpcs ipamTypes.VirtualNetworkMap, sub
 	}
 
 	for _, ip := range iface.PrivateIpAddresses {
-		if ip.PrivateIpAddress != nil {
-			eni.Addresses = append(eni.Addresses, *ip.PrivateIpAddress)
+		if ip.PrivateIpAddress != nil && !ip.Primary {
```
Comment (gandro) [inline pkg/aws/ec2/ec2.go]: This will not compile on 1.8. Please use the following:

```suggestion
		if ip.PrivateIpAddress != nil && (ip.Primary == nil || !*ip.Primary) {
```

File: pkg/aws/ec2/ec2.go [hunk 7]
```diff
@@ -195,8 +195,8 @@ func parseENI(iface *ec2.NetworkInterface, vpcs ipamTypes.VirtualNetworkMap, sub
...
+			eni.Addresses = append(eni.Addresses, aws.ToString(ip.PrivateIpAddress))
```
Comment (gandro) [inline pkg/aws/ec2/ec2.go]: This did not change in the upstream patch. Please revert the change to this line.
```suggestion
			eni.Addresses = append(eni.Addresses, *ip.PrivateIpAddress)
```

File: pkg/logging/logfields/logfields.go [hunk 8]
```diff
@@ -466,4 +475,31 @@ const (
...
+
+	// PodCIDRv6 is the pod IPv4 allocation CIDR
+	PodCIDRv6 = "podCIDRv6"
+
+	// WgIPv4 is the Wireguard IPv4 tunnel addr
+	WgIPv4 = "wgIPv4"
+
+	// WgIPv6 is the Wireguard IPv4 tunnel addr
+	WgIPv6 = "wgIPv6"
+
```
Comment (gandro) [inline pkg/logging/logfields/logfields.go]: I believe all these lines have gotten pulled in accidentally. They must not be backported.

**Review by gandro — CHANGES_REQUESTED** [review]

File: pkg/logging/logfields/logfields.go (hunk 8 above)
Comment (errordeveloper) [reply pkg/logging/logfields/logfields.go]: It was part of https://github.com/cilium/cilium/pull/15453, was that a mistake in the first place?

**Comment by errordeveloper** [comment]
@brb why not add #15261 to your existing backport PR?

**Comment by errordeveloper** [comment]
@joestringer since your changes were quite simple, I've made the changes directly - PTAL!

File: pkg/logging/logfields/logfields.go (hunk 8 above)
Comment (gandro) [reply pkg/logging/logfields/logfields.go]: The ones I highlighted are definitely _not_ part of #15453. The only log fields added via #15453 are `MACAddr`, `Resource`, `Rule`, `AttachedENIs` and `ExpectedENIs`.

Comment (errordeveloper) [reply pkg/logging/logfields/logfields.go]: Ok, I fixed this now.

**Review by gandro — APPROVED** [review]
My PR looks good now.

File: Documentation/gettingstarted/cni-chaining-azure-cni.rst (hunk 5 above)
Comment (joestringer) [reply Documentation/gettingstarted/cni-chaining-azure-cni.rst]: There's two things I see here:
1. Is the author notified that their PR is being backported?
1. How does the author evaluate that the change makes sense on this branch?

My feedback above is intended to address (2). If the cherry-pick is smooth and CI is passing then in all honesty I am probably not going to pay a lot of attention unless I happen to know ahead of time that the backport is error-prone. I already prepared the original change and there was no problem with cherrypick, what useful feedback can I provide on my own change?

If instead, it turns out that the backport required resolving conflicts with the older branch, then it's very useful to know this so I can pay closer attention to the area that the backporter highlighted, to double-check that the changes were equivalent on this branch. In the end, the workflow I would use in this case is to pull up the two commits side-by-side (the original and the backport) and check line-by-line that the changes are equivalent. Strictly speaking I think the backporter can (and should!) do this, but the backporter may not have the context to understand the implications of any differences in the code between the branches. That's why I think original author review can catch things that might otherwise be missed, and actually get value out of the review process for backports.

(Marked unresolved just to help folks see this because we should come to an agreement on these principles as a community so we all understand the expectations around how to communicate clearly on what happened and what we expect from authors of PRs that are backported)

**Review by joestringer — APPROVED** [review]
Docs changes LGTM now, thanks.

File: Documentation/gettingstarted/cni-chaining-azure-cni.rst (hunk 5 above)
Comment (errordeveloper) [reply Documentation/gettingstarted/cni-chaining-azure-cni.rst]: @joestringer thanks, this is definitely very helpful. I was doing this a bit too late in the day, and happy to face too many unlucky commits...

**Comment by errordeveloper** [comment]
still waiting on https://github.com/cilium/cilium/pull/15537

**Comment by errordeveloper** [comment]
test-backport-1.8

**Comment by errordeveloper** [comment]
these ones seem to have failed, will re-test:

k8s-1.14-kernel-4.9 (test-1.14-4.9)
k8s-1.15-kernel-4.9 (test-1.15-4.9)
k8s-1.16-kernel-4.9 (test-1.16-4.9)
k8s-1.18-kernel-4.9 (test-1.18-4.9)

**Comment by errordeveloper** [comment]
test-backport-1.8

**Comment by errordeveloper** [comment]
And on the last attempt 5 jobs failed (last time it was only 4)...

k8s-1.14-kernel-4.9 (test-1.14-4.9)
k8s-1.15-kernel-4.9 (test-1.15-4.9)
k8s-1.16-kernel-4.9 (test-1.16-4.9)
k8s-1.17-kernel-4.19 (test-1.17-4.19)
k8s-1.18-kernel-4.9 (test-1.18-4.9)

**Comment by errordeveloper** [comment]
test-backport-1.8

**Comment by pchaigno** [comment]
@errordeveloper Did you have a look at the failing tests? Why are they failing?

Also, note you can restart individual tests with the trigger phrase indicated in the job's name. That avoids wasting CI resources on full end-to-end tests when many jobs are already passing.

**Comment by errordeveloper** [comment]
@pchaigno thanks, I didn't know that what's in brackets is the trigger-phrase, I'll try it next time.

**Comment by errordeveloper** [comment]
@brb looks like this PR still needs your approval...

**Review by brb — APPROVED** [review]
My changes LGTM, thanks!

**Comment by errordeveloper** [comment]
k8s-1.18-kernel-4.9:

Suite-k8s-1.18.K8sChaosTest Restart with long lived connections TCP connection is not dropped when cilium restarts
Suite-k8s-1.18.K8sChaosTest Restart with long lived connections L3/L4 policies still work while Cilium is restarted

k8s-1.17-kernel-4.19:

Suite-k8s-1.17.K8sVerifier Runs the kernel verifier against Cilium's BPF datapath
Suite-k8s-1.17.K8sFQDNTest Validate that multiple specs are working correctly

**Comment by errordeveloper** [comment]
test-1.18-4.9

**Comment by errordeveloper** [comment]
test-1.17-4.19

**Comment by errordeveloper** [comment]
test-1.16-4.9

**Comment by errordeveloper** [comment]
test-1.14-4.9

**Comment by pchaigno** [comment]
> Suite-k8s-1.17.K8sVerifier Runs the kernel verifier against Cilium's BPF datapath

@errordeveloper Did you look at the error message? That test doesn't flake often, so probably a legit failure.

**Comment by qmonnet** [comment]
test-1.17-4.19

**Comment by aanm** [comment]
test-backport-1.8

**Comment by aanm** [comment]
test-1.15-4.9
